### PR TITLE
Light estimation

### PR DESCRIPTION
**Pull request by brianzinn**
initial support for light estimation
https://github.com/BabylonJS/Babylon.js/issues/8904

### CODE REVIEW

**Comment by brianzinn** [comment]
Able to retrieve primary light direction/intensity as well as ReflectionCubeMap texture.
image

File: src/XR/features/WebXRLightEstimation.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * An estimated cube map representing the users environment.
+     */
+    public get reflectionCubeMap(): Nullable<WebGLTexture> {
+        if (this._xrLightProbe === null) {
+            return null;
+        }
+
+        if (this._reflectionCubeMap = null) {
```
Comment (RaananW) [inline src/XR/features/WebXRLightEstimation.ts:110]: I assume you meant `===`?

**Review by RaananW — COMMENTED** [review]
We want to keep all async operations using observables or promises.

Whenever possible we should expose an updated/added/created observable, so that devs won't need to check for null (and keep on calling until it is ready). 

Another thing - I understand this is initial support, but this requires the dev to apply the light data to a light created by themselves. The options should allow the user to provide a light that will be updated, or ask to create a light to which the light data be applied.
Not sure if this is out of the scope of this PR.

File: src/XR/features/WebXRLightEstimation.ts (hunk 1 above)
Comment (brianzinn) [reply src/XR/features/WebXRLightEstimation.ts:110]: 🤦

File: src/XR/features/WebXRLightEstimation.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            this._reflectionCubeMap = this._getXRGLBinding().getReflectionCubeMap(this._xrLightProbe);
+            this._xrLightProbe.addEventListener('reflectionchange', this._updateReflectionCubeMap);
```
[the remark this answers is not in the document]
Comment (brianzinn) [reply src/XR/features/WebXRLightEstimation.ts:112]: getting the cubemap is not async (it's only getting the light probe that may take multiple frames, but on my device is available on first frame).  The event listener should actually be a separate feature option.   I'll add that change in and make it clearer.

**Comment by brianzinn** [comment]
I added an observable for the Reflection Cube Map.  It is updated every 20-30 frames on my device.  I suspect you would want light direction and intensity to be `Vector3` for public API (Looks like light intensity needs to be scaled down by the highest value) instead of the Web XR specific typings.  The cube texture as well should likely update a babylon equivalent.  Hope this is enough to get light-estimation feature to next step with spherical harmonics and the WebGLTexture cube map available.

File: src/XR/features/WebXRLightEstimation.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            this._xrWebGLBinding = new XRWebGLBinding(this._xrSessionManager.session, context);
+        }
+        return this._xrWebGLBinding;
+    }
+
+    /**
+     * Event Listener to for "reflectionchange" events.
+     */
+    private _updateReflectionCubeMap = (): void => {
+        this._reflectionCubeMap = this._getXRGLBinding().getReflectionCubeMap(this._xrLightProbe!);
```
Comment (RaananW) [inline src/XR/features/WebXRLightEstimation.ts:108]: does the old one needs to be disposed?

File: src/XR/features/WebXRLightEstimation.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        if (!super.attach()) {
+            return false;
+        }
+
+        this._xrSessionManager.session.requestLightProbe({
+            reflectionFormat: this.options.reflectionFormatSRGBA8 ? "srgba8" : this._xrSessionManager.session.preferredReflectionFormat ?? "srgba8"
+        }).then((value: XRLightProbe) => {
+            this._xrLightProbe = value;
+            this._xrLightProbe.addEventListener('reflectionchange', this._updateReflectionCubeMap);
+            this._reflectionCubeMap = this._getXRGLBinding().getReflectionCubeMap(this._xrLightProbe!);
```
Comment (RaananW) [inline src/XR/features/WebXRLightEstimation.ts:128]: should probably notify here (the observable)

File: src/XR/features/WebXRLightEstimation.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+     * XR Light Probe also contains the probe space, may update it's pose over time as the user moves around their environment.
+     */
+    public get xrLightProbe(): Nullable<XRLightProbe> {
+        return this._xrLightProbe;
+    }
+
+    /**
+     * The most recent light estimate.  Available starting on the first frame where the device provides a light probe.
+     */
+    public get xrLightingEstimate(): Nullable<XRLightEstimate> {
```
Comment (RaananW) [inline src/XR/features/WebXRLightEstimation.ts:87]: we'll need to convert those values to left-handed when used in babylon. otherwise they are flipped.

**Review by RaananW — COMMENTED** [review]
Hey @brianzinn , thank you so much for working on that!

I added a few comments - but I will be happy to work on that if you can't find the time.

File: src/XR/features/WebXRLightEstimation.ts (hunk 3 above)
Comment (brianzinn) [reply src/XR/features/WebXRLightEstimation.ts:108]: @RaananW  I didn't clean up based on this:
https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.8
```
Note that underlying GL object will be automatically marked for deletion when the JS object is destroyed, however this method allows authors to mark an object for deletion early.
```

It's easy for me to mark for deletion  - I started making some changes to do so.  Shall I update with `context.deleteTexture(this._reflectionCubeMap)`?

File: src/XR/features/WebXRLightEstimation.ts (hunk 4 above)
Comment (brianzinn) [reply src/XR/features/WebXRLightEstimation.ts]: fixed.

File: src/XR/features/WebXRLightEstimation.ts (hunk 5 above)
Comment (brianzinn) [reply src/XR/features/WebXRLightEstimation.ts:88]: Yes @RaananW - To be consistent I think we want to remove XR specific typings from the API.  Likely that means wrapping with something like:
```typescript
export type LightEstimate = {
  primaryLightDirection: Vector3 // changed to left handed
  primaryLightIntensity: Vector3 // although not directly usable (see below)
  sphericalHarmonicCoefficients: Float32Array
}
```
I didn't go that extra step.  Following the explainer the `intensity` from the primary light is used to set color and a numeric primitive for light intensity (https://github.com/immersive-web/lighting-estimation/blob/main/lighting-estimation-explainer.md#xrlightestimate).

**Comment by RaananW** [comment]
Note - I will merge this right after making a few small changes (i.e. the last discussion). Will keep this open until then.

**Comment by sebavan** [comment]
Thanks @RaananW I ll leave it open for you then :-)

**Review by sebavan — CHANGES_REQUESTED** [review]
Preventing merge before @RaananW can jump in :-)

**Comment by sebavan** [comment]
@brianzinn did you have any chance to take a look at the changes ?

**Comment by brianzinn** [comment]
> @brianzinn did you have any chance to take a look at the changes ?

@sebavan I have not been looking at the changes - I am under the impression that @RaananW was going to independently make changes and then to merge.  If there is something specific for me to follow up on then just let me know and I am happy to look into.  Thanks.

**Comment by sebavan** [comment]
Oh my.... completely misread the thread and the worst part is I already discussed it with him.... so yes nothing required from your side for now :-)

**Comment by sebavan** [comment]
Closing as @RaananW is implementing it :-)

**Pull request closed without merging**